python/ray/data/_internal/logical/optimizers.py

```diff
@@ -0,0 +1,48 @@
+from typing import List
+
+from ray.data._internal.execution.interfaces import PhysicalOperator
+from ray.data._internal.logical.interfaces import Rule, Optimizer, LogicalOperator
+
+
+class LogicalOptimizer(Optimizer):
+    """The optimizer for logical operators."""
+
+    @property
+    def rules(self) -> List[Rule]:
+        # TODO: Add logical optimizer rules.
+        return []
+
+
+class PhysicalOptimizer(Optimizer):
+    """The optimizer for physical operators."""
+
+    @property
+    def rules(self) -> List["Rule"]:
+        # TODO: Add physical optimizer rules.
+        return []
+
+
+class LogicalPlan:
+    """The plan with a DAG of logical operators."""
+
+    def __init__(self, dag: LogicalOperator):
+        self._dag = dag
+
+    def get_new_plan(self, operator: LogicalOperator) -> "LogicalPlan":
```

Comment:
This seems only going to produce linear plan as the current Stages, shall we need more generic interface?